### PR TITLE
Gamepad support

### PR DESCRIPTION
**Pull request by wiktorek140**
Currently fully suported is Gamepad Tracer Shogun USB.
Config file should be in game data main folder and named gamepad.conf
I take custom lib to read input event and configure it to gamepad.
To support other devices i need keymap made by test program from https://github.com/drewnoakes/joystick in format "[key number] - coresponding key function (A/B and other)"

### CODE REVIEW

**Comment by wiktorek140** [comment]
oh, and i forgot one thing, if pad has Dpads as button not axis it need to rewrite some part of code. config file must look like these i provide

File: src/gamepad.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    int DPAD_LR,
+        DPAD_UD;
+
+    struct StickValueInfo {
+        float x = 0.f, y = 0.f;
+    };
+
+    //int getFreeDeviceIndex();
+
+    //void onDeviceAdded(struct udev_device* dev);
```
Comment (InusualZ) [inline src/gamepad.h]: Delete this unused declaration

File: src/gamepad.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                    if(event.value > 0) lastdpad[0] = BUTTON_DPAD_RIGHT;
+                    if(event.value < 0) lastdpad[0] = BUTTON_DPAD_LEFT;
+                    GameControllerManager::sGamePadManager->feedButton(0, lastdpad[0], event.value !=0 ? 1 : 0, true);
+                }
+                if(event.number == DPAD_UD) { //dpad usage
+                    if(event.value > 0) lastdpad[1] = BUTTON_DPAD_DOWN;
+                    if(event.value < 0) lastdpad[1] = BUTTON_DPAD_UP;
+                    GameControllerManager::sGamePadManager->feedButton(0, lastdpad[1], event.value !=0 ? 1 : 0, true);
+                }
+
```
Comment (InusualZ) [inline src/gamepad.cpp]: There is anyway to configure the axis? I don't think Minecraft handle that right?

Comment (wiktorek140) [reply src/gamepad.cpp]: You can reconfigure it there but not in game

**Comment by MCMrARM** [comment]
Not doing a full code review currently as first I want to do a few comments on this.

First of all, this uses the deprecated joystick API. If I were to stick to it, I'd definitely want to have full SDL mappings support. Otherwise, I implemented gamepad support In the `gamepad-support` branch using the newer evdev API and if I were to add a remapping tool I'd go with that one instead with that (mine) solution.

Another issue with your implementation is that always the joystick 0 is opened with a hardcoded path. An user might want to use more than one gamepad. This also is missing an autodetection API.

The biggest issue I had with my approach I used is that it wasn't compatible with any existing mappings and the mappings would have to be redone (evdev uses a completely different mapping for whoever-knows-the-reason).

**Comment by wiktorek140** [comment]
So if i add support to read config from file and make mapping it will be better?

**Comment by MCMrARM** [comment]
May I know why did you ditch the current `gamepad-support` and start from scratch?

**Comment by wiktorek140** [comment]
Because i cannot configure buttons in these solution. And event was quite unreadable for me

**Comment by MCMrARM** [comment]
Well, it shouldn't be too hard to implement an HTML-based gamepad mapper.

**Comment by MCMrARM** [comment]
I currently would prefer to stick to my solution and implement a mapper instead.

**Comment by wiktorek140** [comment]
So i will try to do it in your way ;)

**Comment by MCMrARM** [comment]
I'll work on the gamepad mapper in the `gamepad-support` branch now.

**Comment by MCMrARM** [comment]
Some work on the gamepad mapping was done on the `gamepad-support` (it is not yet saved, but the basic UI is done).

**Pull request closed without merging**